Prject Service -BJS2-40695 (sprint 6)

Amanyaz opened the pull request:
Это задача BJS2-40695 (БЕЗ ТЕСТА) из спринта 6. Требования не такие ч…еткие, как в спринте 5. Поэтому я сделал, как понял.

Amanyaz committed:
Это задача BJS2-40695 (БЕЗ ТЕСТА) из спринта 6. Требования не такие четкие, как в спринте 5. Поэтому я сделал, как понял.

Amanyaz committed:
Это задача BJS2-40695 (С ТЕСТАМИ) из спринта 6. Требования не такие четкие, как в спринте 5. Поэтому я сделал, как понял.

RuslanGabd commented on src/main/java/faang/school/projectservice/service/MomentService.java:
```diff
@@ -0,0 +1,60 @@
+    private final List<MomentFilter> momentFilters;
+
+    public MomentDto create(MomentDto momentDto) {
+        Moment moment = momentMapper.toEntity(momentDto);
```
я бы добавил проверку что бы убедиться что такого Момента еще нету в базе и по идеи мы не должны id в Dto передавать, id БД выдает в процессе записи сущности. Фронтенд точно не формиурет id сущности

Amanyaz replied:
Я удалил поле id из класса MomentDto. Но я не совсем понял, как проверить, создан ли такой момент в базе данных или нет. Насколько я знаю, если момент уже существует, база данных просто обновит его, а если момента нет, то создаст его. Если у вас есть другая идея, пожалуйста, исправьте меня и буду признателен, если вы предоставите необходимый код.

Amanyaz replied:
Я удалил поле id из класса MomentDto.  я создал метод для проверки, существует ли момент. Если существует, он выбрасывает исключение, если не существует — создаёт его.

RuslanGabd replied:
здесь все правильно

RuslanGabd commented on src/main/java/faang/school/projectservice/service/MomentService.java:
```diff
@@ -0,0 +1,60 @@
+            return momentMapper.toDto(foundMoment);
+        }
+        throw new ValidationException("There is no moment with given Id");
+    }
```
в следующей можно попробовать подключить логи, что бы если была ошибка она еще и в логи писалась.
Как пример: 
![image](https://github.com/user-attachments/assets/21c84cb5-3f8f-4eec-b8f5-2b49064fa901)

RuslanGabd commented on src/main/java/faang/school/projectservice/dto/moment/MomentDto.java:
```diff
@@ -0,0 +1,24 @@
+    private String name;
+    private String description;
+    private List<Resource> resource;
+    private List<Project> projects;
```
не должно дто содержать другие сущности темболее списки. Максимум id других сущностей/ списки id 
Иначе представляешь какой трафик будет если между беком и фронтом будет передаваться обьект со списсками других обьектов

Amanyaz replied:
Да, вы правы, Влад тоже говорил об этом во время своего видеоурока, но я забыл. Поэтому я изменил код так, как Влад показывал на уроке: удалил List<Resource> resource, потому что она мне вообще не нужна в классе MomentDto, и заменил List<Project> projects на List<ProjectDto> projects. Я сам создал класс ProjectDto, и в нем только два поля: long id и ProjectStatus status. ProjectStatus — это не сущность, она не имеет представления в базе данных, это просто перечисление (Enumeration).

RuslanGabd replied:
я не вижу изменений кода

RuslanGabd commented on src/main/java/faang/school/projectservice/controller/moment/MomentController.java:
```diff
@@ -0,0 +1,59 @@
+        }
+        if (momentDto.getProjects().stream().anyMatch(project -> project.getStatus()
+                .equals(ProjectStatus.COMPLETED))) {
+            throw new ValidationException("The project was completed");
```
исходя из комментария в MomentDto эти проверки здесь не должны быть, из Дто мы получаем айдишников проджектов через репозатирии проджекта вытаскиваем этот проджект и проверяем, такая проверка должна делаться в сервисе.

Amanyaz replied:
Сделано!

RuslanGabd replied:
> Сделано!

Где?

RuslanGabd reviewed (commented):
все ок, но есть пару замечаний. Dto надо переделывать.

Amanyaz committed:
This commit includes removal of entites as a dependencies of MomentDto, and create method includes validation method isExists() in the MomentService class, if the moment was created or not. Here, I aligned all the tested with above changes, and all the tests passes.

Amanyaz committed:
This commit includes below:
-Validation of the status COMPLETED was moved to MomentService from MomentController.
- Validation of the status CANCELLED was moved to MomentService from MomentController.
- injected dependency of ProjectRepository class to get projects by List<Long> Ids.
- All the tests were aligned to the above alterations, and all the tests passes successfully.

Amanyaz committed:
This commit includes below:
-Validation of the status COMPLETED was moved to MomentService from MomentController.
- Validation of the status CANCELLED was moved to MomentService from MomentController.
- injected dependency of ProjectRepository class to get projects by List<Long> Ids.
- All the tests were aligned to the above alterations, and all the tests passes successfully.
-No logs

jibulani commented on src/main/java/faang/school/projectservice/dto/moment/MomentFilterDto.java:
```diff
@@ -0,0 +1,14 @@
+
+@Data
+public class MomentFilterDto {
+    Project project;
```
дто не должен использовать сущности из базы

Amanyaz replied:
Заменил на ProjectDto

jibulani commented on src/main/java/faang/school/projectservice/dto/moment/filters/MomentFilter.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface MomentFilter {
+boolean isApplicable(MomentFilterDto filter);
+
```
проблемы с отступами

Amanyaz replied:
поправил

jibulani commented on src/main/java/faang/school/projectservice/dto/project/ProjectDto.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+public class ProjectDto {
+long id;
```
и тут отступы уехали

Amanyaz replied:
Поправил

jibulani commented on src/test/java/faang/school/projectservice/service/project/MomentServiceTest.java:
```diff
@@ -0,0 +1,230 @@
+        MomentStartDateFromFilter filterStartDateFromMock = Mockito.mock(MomentStartDateFromFilter.class);
+        momentFilterListMock = List.of(filterPartnerProjectMock, filterStartDateFromMock);
+    }
+
```
договаривались писать названия тестов в формате, при котором в названии будет бизнес-условие и успешность теста, т.е. testCreateWithExistingMoment можно переписать в виде createWithExistingMomentFailTest, а testCreate - createMomentSuccessTest
лучше так именуй, чтобы в проекте все тесты в одинаковом формате назывались

Amanyaz replied:
изменил

jibulani commented on src/main/java/faang/school/projectservice/service/MomentService.java:
```diff
@@ -0,0 +1,91 @@
+
+    public MomentDto update(MomentDto momentDto) {
+
+        Moment moment = momentMapper.toEntity(momentDto);
```
как достигается выполнение требования?

> Если добавляется новый партнёр, то должен обновиться список участников и наоборот: если добавляется участник из другого проекта, то проект автоматически добавляется в партнеры по моменту.

Amanyaz replied:
сделал

jibulani reviewed (changes requested):
хорошо, но есть несколько мест для исправления

Amanyaz committed:
Я обновил метод update() в классе MomentService. Я добавил добавление партнерских проектов и соответствующих членов команды.

jibulani reviewed (approved):
принято, молодец!
